src/main/java/io/anserini/rerank/AxiomReranker.java

```diff
@@ -0,0 +1,302 @@
+        rs = searcher.search(q, 1000);
+      }
+
+      return ScoredDocuments.fromTopDocs(rs, searcher);
+    } catch (Exception e) {
+      System.out.println(e);
+    }
+    return null;
+  }
+
+  /** Select R*M docs from the ranking results and the index as the reranking pool.
+   *  The process is:
+   *  1. Keep the top M documents in the original ranking list
+   *  2. Randomly pick (R-1)*M documents from the rest of the index so in total we have R*M documents
+   *
+   *  @param docs      The initial ranking results
+   *  @param context   An instance of RerankerContext
+   *
+   *  @return a Set of R*M document Ids
+   */
+  @VisibleForTesting
+  private Set<Integer> selectDocs(ScoredDocuments docs, RerankerContext context)
+      throws IOException {
+    Set<Integer> docidSet = new HashSet<>(Arrays.asList(ArrayUtils.toObject(
+        Arrays.copyOfRange(docs.ids, 0, Math.min(this.M, docs.ids.length)))));
+    long targetSize = this.R * this.M;
+
+    if (docidSet.size() < targetSize) {
+      IndexSearcher searcher = context.getIndexSearcher();
+      IndexReader reader = searcher.getIndexReader();
+      int availableDocsCnt = reader.getDocCount(this.field);
+      List<Integer> docidsPoolList = new ArrayList<>();
+      for (int i = 1; i <= availableDocsCnt; i++) {
+        docidsPoolList.add(i);
+      }
+      Random random = new Random();
+      while (docidSet.size() < targetSize) {
+        docidSet.add(docidsPoolList.get(random.nextInt(docidsPoolList.size())));
+      }
+    }
+
+    return docidSet;
+  }
+
+  /** Extract ALL the terms from the documents pool.
+   *
+   *  @param docIds          The reranking pool, see {@link #selectDocs} for explanations
+   *  @param context         An instance of RerankerContext
+   *  @param filterPattern   A Regex pattern that terms are collected only they matches the pattern, could be null
+   *
+   *  @return A Map of <term -> Set<docId>> kind of a small inverted list where the Set of docIds is where the term occurs
+   */
+  @VisibleForTesting
+  private Map<String, Set<Integer>> extractTerms(Set<Integer> docIds, RerankerContext context,
+                                                 Pattern filterPattern) throws Exception, IOException {
+    IndexSearcher searcher = context.getIndexSearcher();
+    IndexReader reader = searcher.getIndexReader();
+    Map<String, Set<Integer>> termDocidSets = new HashMap<>();
+    for (int docid : docIds) {
+      Terms terms = reader.getTermVector(docid, LuceneDocumentGenerator.FIELD_BODY);
+      if (terms == null) {
+        LOG.info("Document vector not stored for docid: " + docid);
+        continue;
+      }
+      TermsEnum te = terms.iterator();
+      if (te == null) {
+        LOG.info("Document vector not stored for docid: " + docid);
+        continue;
+      }
+      while ((te.next()) != null) {
+        String term = te.term().utf8ToString();
+        if (term.length() < 2) continue;
+        if (!term.matches("[a-z]+")) continue;
```

Comment:
This is some type of heuristic I assume? Document it a bit... I assume this cuts down on noisy terms?

Reply:
Yes, will add some comments.